stack-graphs/src/storage.rs

```diff
@@ -0,0 +1,472 @@
+// -*- coding: utf-8 -*-
+// ------------------------------------------------------------------------------------------------
+// Copyright © 2023, stack-graphs authors.
+// Licensed under either of Apache License, Version 2.0, or MIT license, at your option.
+// Please see the LICENSE-APACHE or LICENSE-MIT files in this distribution for license details.
+// ------------------------------------------------------------------------------------------------
+
+use rusqlite::Connection;
+use std::collections::HashSet;
+use std::path::Path;
+use thiserror::Error;
+
+use crate::arena::Handle;
+use crate::graph::File;
+use crate::graph::Node;
+use crate::graph::StackGraph;
+use crate::partial::PartialPath;
+use crate::partial::PartialPaths;
+use crate::partial::PartialSymbolStack;
+use crate::serde;
+use crate::serde::FileFilter;
+use crate::stitching::Database;
+use crate::stitching::ForwardPartialPathStitcher;
+use crate::CancellationError;
+use crate::CancellationFlag;
+
+const VERSION: usize = 1;
+
+const SCHEMA: &str = r#"
+        CREATE TABLE metadata (
+            version INTEGER NOT NULL
+        ) STRICT;
+        CREATE TABLE graphs (
+            file TEXT PRIMARY KEY,
+            tag  TEXT NOT NULL,
+            json BLOB NOT NULL
+        ) STRICT;
+        CREATE TABLE file_paths (
+            file     TEXT NOT NULL,
+            local_id INTEGER NOT NULL,
+            json     BLOB NOT NULL,
+            FOREIGN KEY(file) REFERENCES graphs(file)
+        ) STRICT;
+        CREATE TABLE root_paths (
+            file         TEXT NOT NULL,
+            symbol_stack TEXT NOT NULL,
+            json         BLOB NOT NULL,
+            FOREIGN KEY(file) REFERENCES graphs(file)
+        ) STRICT;
+    "#;
+
+#[derive(Debug, Error)]
+pub enum StorageError {
+    #[error("cancelled at {0}")]
+    Cancelled(&'static str),
+    #[error("unsupported database version {0}")]
+    IncorrectVersion(usize),
+    #[error("database does not exist {0}")]
+    MissingDatabase(String),
+    #[error(transparent)]
+    Rusqlite(#[from] rusqlite::Error),
+    #[error(transparent)]
+    Serde(#[from] serde::Error),
+    #[error(transparent)]
+    SerdeJson(#[from] serde_json::Error),
+}
+
+pub type Result<T> = std::result::Result<T, StorageError>;
+
+impl From<CancellationError> for StorageError {
+    fn from(value: CancellationError) -> Self {
+        Self::Cancelled(value.0)
+    }
+}
+
+/// Writer to store stack graphs and partial paths in a SQLite database.
+pub struct SQLiteWriter {
+    conn: Connection,
+}
+
+impl SQLiteWriter {
+    /// Open an in-memory database.
+    pub fn open_in_memory() -> Result<Self> {
+        let conn = Connection::open_in_memory()?;
+        Self::init(&conn)?;
+        Ok(Self { conn })
+    }
+
+    /// Open a file database.  If the file does not exist, it is automatically created.
+    /// An error is thrown if the database version is not supported.
```

Comment:
Small nit: Could this be updated to say `An error is returned ...` instead of `throws`?